bionic/code_analyzer.py

```diff
@@ -0,0 +1,181 @@
+                self._hash.update(CodeHashPrefix.INT.value)
+            elif isinstance(obj, float):
+                self._hash.update(CodeHashPrefix.FLOAT.value)
+            elif isinstance(obj, str):
+                self._hash.update(CodeHashPrefix.STRING.value)
+            elif isinstance(obj, bool):
+                self._hash.update(CodeHashPrefix.BOOL.value)
+            json_bytes = json.dumps(obj).encode()
+            self._hash.update(json_bytes)
+            self._hash.update(get_size_in_bytes(json_bytes))
+
+        elif isinstance(obj, (list, tuple)):
+            if isinstance(obj, list):
+                self._hash.update(CodeHashPrefix.LIST.value)
+            elif isinstance(obj, tuple):
+                self._hash.update(CodeHashPrefix.TUPLE.value)
+            for elem in obj:
+                self._update_with_checks(elem, depth + 1)
+
+        elif isinstance(obj, dict):
+            self._hash.update(CodeHashPrefix.DICT.value)
+            for key, elem in obj.items():
+                self._update_with_checks(key, depth + 1)
+                self._update_with_checks(elem, depth + 1)
+
+        elif inspect.isroutine(obj):
+            self._hash.update(CodeHashPrefix.ROUTINE.value)
+            if obj.__defaults__ is not None:
+                for d in obj.__defaults__:
+                    self._update_with_checks(d, depth + 1)
+
+            self._analyze_code(obj.__code__, depth)
+
+        elif inspect.iscode(obj):
+            self._hash.update(CodeHashPrefix.CODE.value)
+            self._analyze_code(obj, depth)
+
+        else:
+            self._hash.update(CodeHashPrefix.DEFAULT.value)
+            message = oneline(
+                f"""
+                Found a constant {obj!r} of type {type(obj)!r} that
+                Bionic cannot hash. This is most likely a miss from the
+                developers. Please contact one of the code owners of
+                Bionic to fix this issue.
+                """
+            )
+            warnings.warn(message)
+
+    def _analyze_code(self, code, depth):
+        # TODO: Find references for the code and analyze references.
+        self._update_with_checks(code.co_code, depth + 1)
+
+        # Hash constants used in the code object, but ignore names of lambdas.
+        consts = [
+            const
+            for const in code.co_consts
+            if not isinstance(const, str) or not const.endswith(".<lambda>")
+        ]
+        self._update_with_checks(consts, depth + 1)
+
+
+class CodeHashPrefix(Enum):
+    """
+    Represents a unique value for each type that CodeHasher hashes that
+    is prefixed to avoid collision between same encoded values.
+
+    If you change the prefix of any type, or add a new prefix, it can
+    change the encoding of objects. Since Bionic use to encoding makes
+    caching backwards-incompatible.
+
+    Update CACHE_SCHEMA_VERSION to update cache scheme if you change
+    the encoding.
+    """
+
+    BYTES = b"AA"
```

Comment:
They don't have to be. In case we add more than 26 cases and start using `ZA`, `ZB`, and so on, I think it'd be nice to have the prefix's length consistent.